utils/travis-helper.sh

```diff
@@ -0,0 +1,475 @@
+#
+# Basic Idea:
+# ------------
+# So the idea is to build painful stack incl. ghc only once in the 1st stage
+# named 'Stack' and use it as is for all subsequent stages and builds. This
+# produces ~ 2.8 GiB bloat (uncompressed) => ~ 825 MiB tgz cache archive.
+#
+# In the next stage we build hets (desktop version) and hets_server in parallel
+# using 2 jobs. Basic idea is to not use ${TRAVIS_BUILD_DIR} which is, what
+# travis-ci checks out from github and gets used per default for build. Instead
+# we sync the cached dir ${HETS_BASEDIR}/{desktop|server} with it using rsync
+# and its size+checksum algorithm. However, because the dumb stack stuff hard-
+# codes pathes into its package DBs, meta data files, we can't use the easy way
+# (2 dirs directly), but have to rename ${HETS_BASEDIR}/{desktop|server} to
+# e.g. ${HETS_BASEDIR}/shared and rename it back, after the job is done.
+# Unfortunately this is still not sufficient because of travis' dumb caching:
+# the job which finishes 1st would package and x-fer the new cache archive
+# (takes ~ 3-4 min) to google and later, when the 2nd job gets finished it does
+# the same overwriting the "cache" from the 1st job. So we need to use external
+# caching at least for the result of one job - best candidate is server, because
+# slightly smaller and used by one test only (all other tests use the hets
+# desktop version). Archiving ${HETS_BASEDIR}/{desktop|server} takes about
+# 20+-5s depending on the machine and whether the final binary gets included,
+# and produces a tgz of ~ 100 MB (+ 21 MB if the binary gets included). Xfer
+# rate is ~ 75+-20 MiB/s so add. costs incl. extracting would be ~ 3..6s.
+# Compared to the other cached stuff this is relatively cheap and thus we
+# simply synchronize the desktop 'Build' job with the server 'Build' job via
+# our own external cache, and pull in the "server archive cache", before the
+# desktop job creates and uploads its new cache.
+# A little bit better would be to always push and pull on demand the
+# hets[_server] archives to the external cache, only. However, this would
+# require more know-how, maintenance burden, complexity on the cacher site ...
+#
+# In the last stage named 'Test' we run all test jobs in parallel. We just need
+# to take care of properly renaming the appropriate
+# ${HETS_BASEDIR}/{desktop|server} to ${HETS_BASEDIR}/shared - same way as in
+# the 'Build' stage. Because we do not need any changes we explicitly disable
+# cache uploads by removing the related script: ${CASHER_DIR}/bin/casher
+
+# Usable Travis env vars
+# ======================
+# TRAVIS_BUILD_ID=396668280
+# TRAVIS_BUILD_NUMBER=2016
+# TRAVIS_BUILD_DIR=/home/travis/build/spechub/Hets
+# TRAVIS_BRANCH=master
+# TRAVIS_EVENT_TYPE=pull_request
+# TRAVIS_PULL_REQUEST=1853
+# TRAVIS_PULL_REQUEST_BRANCH=dependency_fix
+# TRAVIS_JOB_ID=396668281
+# TRAVIS_JOB_NUMBER=2016.1
+
+[[ -z "${PUSH_URL}" ]] && PUSH_URL='http://theo.cs.ovgu.de/cgi-bin/push'
```

Comment:
Can this be abused to upload (and download) arbitrary data to the server in any way? Also, will the external cache automatically cleaned after a set amount of time, or is this a manual task?

Reply:
Basically yes. However, to prevent DoS, the upload directory is a separate filesystem. For upload the parameters have to match certain patterns (little bit "security by obscurity" ;-)), and uploads/downloads are restricted to IKS intern and travis-CI sites, only. Last but not least, directory indexing is disabled, so one needs to know, what to download.

So with some effort and proper timing, cache poisoning is possible. However, I think it is unlikely. If it becomes a problem, stuff can be probably enhanced by using Travis-CI encryption stuff (haven't had the time to study this stuff).